### PR TITLE
Add Common Repository Structure

### PR DESCRIPTION
**Pull request by shemnon**
Add a narrative about the common repository structure

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

### CODE REVIEW

**Comment by shemnon** [comment]
This pulls over the content from https://github.com/hyperledger-labs/hyperledger-community-management-tools/blob/main/repo_structure/README.md as a TSC document.

File: repository-structure.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        [Apache License, Version 2](https://www.apache.org/licenses/LICENSE-2.0)
+-   Apache License Header information in each source code file. \
+    This SHOULD include the snippet `SPDX-License-Identifier: Apache-2.0` as part of the header.
+-   Build files consistent with the implementation language, such as...
+    -   For JavaScript/Node.js a `package.json` file
+    -   For Ruby a `Gemfile` file
+    -   For Java one of a Maven `pom.xml`, an Apache Ant `build.xml`, or a Gralde `build.gradle`
+        file
+    -   For Python `setup.py` and `requirements.txt` files
+    -   For Go `go.mod`, `go.sum` and as an option a Makefile
```
Comment (arsulegai) [inline repository-structure.md]: Q: Can `go.sum` be an optional file as well?

File: repository-structure.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    A description of the project and contain information or links to information such as
+    -   A reference to the Apache license (required).
+    -   The current and important past releases
+    -   Documentation for developers and users
+-   `MAINTAINERS` \
+    A list of all current maintainers with contact info. [A separate document](MAINTAINERS-guidelines.md)
+    covers the specifics.
+-   `CONTRIBUTING` \
+    Directions on how to contribute code to the project, or a pointer to the Wiki page with that information.
+-   `CHANGELOG` \
```
Comment (arsulegai) [inline repository-structure.md:32]: `Hyperledger Sawtooth` project requested that `CHANGELOG` file be optional. The project follows elaborate release process which includes creating a tag, release notes and the commit history is pretty clean. Please refer to https://github.com/hyperledger-labs/hyperledger-community-management-tools/pull/67

File: repository-structure.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+Hyperledger projects are required to maintain a standard set of files in each repository. This
+document describes the required and recommended files.
+
+## Required Files with Specified Content
+
+Repositories MUST have these files with the specific content in the linked files, or a file with a
+link to the specified content with minimal exposition. These files MUST be at the root of the
+repository.
+
+-   [`LICENSE`](https://www.apache.org/licenses/LICENSE-2.0.txt) - https://www.apache.org/licenses/LICENSE-2.0.txt
```
Comment (arsulegai) [inline repository-structure.md]: There can be a repository which is not `Apache 2.0`. For example, an exception as per the TSC charter or `Creative Commons` for docs. Is it better to redirect readers to the appropriate locations elsewhere to read elaborately on allowed licenses? However, a `LICENSE` file can be mandated here.

File: repository-structure.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+## Recommended
+
+Repositories SHOULD have these files. Named files SHOULD be at the root of the repository
+
+-   `NOTICE`
+    -   As per section 4 subsection d of the
+        [Apache License, Version 2](https://www.apache.org/licenses/LICENSE-2.0)
+-   Apache License Header information in each source code file. \
+    This SHOULD include the snippet `SPDX-License-Identifier: Apache-2.0` as part of the header.
```
Comment (arsulegai) [inline repository-structure.md]: [nitpick] Either the snippet `SPDX..` or text as defined in the Apache 2.0 license.

File: repository-structure.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    -   For Rust a `cargo.toml` file
+    -   For other languages, other standard build files a practitioner of the language would expect.
+-   Testing code \
+    Code to test the code in the repository (such as unit tests), in a location appropriate for the language.
+    \
+    Not all repositories can be tested (homebrew, docs), which is the only reason this is a SHOULD.
+
+## Prohibited
+
+Repositories MUST NOT have these files
```
Comment (arsulegai) [inline repository-structure.md:65]: @hyperledger/tsc is there a defined process for releasing/publishing built artefacts? If not, maybe we can/should take a look at the current process and come up with recommendation.

**Review by arsulegai — CHANGES_REQUESTED** [review]

File: repository-structure.md (hunk 1 above)
Comment (shemnon) [reply repository-structure.md]: Sure.

File: repository-structure.md (hunk 5 above)
Comment (shemnon) [reply repository-structure.md:65]: No.  We should form a task force, IMHO it's a problem.

File: repository-structure.md (hunk 4 above)
Comment (shemnon) [reply repository-structure.md]: https://wiki.hyperledger.org/display/TSC/Copyright+and+License+Policy requires the SPDX form.

File: repository-structure.md (hunk 3 above)
Comment (shemnon) [reply repository-structure.md]: This requires an act of the governing board.  Added in that the exception may occur.

File: repository-structure.md (hunk 2 above)
Comment (shemnon) [reply repository-structure.md:32]: I think that's too substantial a change to move with only two TSC members reviewing a PR.  When the TSC as a whole voted on it it was included as required.

My take is that in this case the CHANGELOG could stay and point to the Github repository changes list.  However this reduces the utility of the file when the repo is cloned by a non HLP project.

File: repository-structure.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+## Required Files with Specified Content
+
+Repositories MUST have these files with the specific content in the linked files, or a file with a
+link to the specified content with minimal exposition. These files MUST be at the root of the
+repository.
+
+-   [`LICENSE`](https://www.apache.org/licenses/LICENSE-2.0.txt) - https://www.apache.org/licenses/LICENSE-2.0.txt\
+    (Unless an exception has been made by the Hyperledger Governing Board)
+-   [`CODE_OF_CONDUCT.md`](https://wiki.hyperledger.org/community/hyperledger-project-code-of-conduct) - https://wiki.hyperledger.org/community/hyperledger-project-code-of-conduct
```
Comment (tkuhrt) [inline repository-structure.md]: Should we be pointing to https://tsc.hyperledger.org/code-of-conduct.html as the official document?

File: repository-structure.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+-   `MAINTAINERS` \
+    A list of all current maintainers with contact info. [A separate document](MAINTAINERS-guidelines.md)
+    covers the specifics.
+-   `CONTRIBUTING` \
+    Directions on how to contribute code to the project, or a pointer to the Wiki page with that information.
+-   `CHANGELOG` \
+    A human readable list of recent changes. Changes should at least include the current release. This
+    file may be maintainer curated or mechanically produced.
+-   Continuous Integration / Continuous Delivery (CICD) configurations \
+    Configurations needed to run CICD on Hyperledger provided systems.
```
Comment (tkuhrt) [inline repository-structure.md:36]: Do we still have Hyperledger provided CICD systems?

Comment (shemnon) [reply repository-structure.md:36]: Github actions is paid for by hyperledger, as is CircleCI. So the intent (allowing for future changes) is to say run it as github actions or on circlci. That way if jenkins came back (it won't), or drone.io or some other CI provider was paid for the TSC wouldn't have to amend this document.

Is there different verbiage that should be used?

File: repository-structure.md (hunk 6 above)
Comment (shemnon) [reply repository-structure.md]: Done. I take it this will be the new cannonical home of the CoC doc?

File: repository-structure.md (hunk 7 above)
Comment (ryjones) [reply repository-structure.md:36]: To be clear, GitHub actions are free; right now, Hyperledger pays for CircleCI and Azure Pipelines.

This is a huge issue, as several projects run in opaque Jenkins silos where inspection of build errors is not possible without an account.

File: repository-structure.md (hunk 2 above)
Comment (arsulegai) [reply repository-structure.md:32]: Happy to include this as a subject for amendment to the earlier proposal.

File: repository-structure.md (hunk 4 above)
Comment (arsulegai) [reply repository-structure.md]: Wiki recommends `SPDX` form. Then there is a section that says

```
IMPORTANT: The license of a file, regardless of how specified, should not be modified unless agreed to by all contributors to the file.
```

There are projects which still follow the text as defined in the Apache 2.0 license. If we are mandating the format then it shall go through a round of TSC discussion.

File: repository-structure.md (hunk 7 above)
Comment (tkuhrt) [reply repository-structure.md:36]: Good to know. I knew that things were being pushed towards GitHub actions. I did not realize there were still CircleCI and Azure Pipelines being used. I was commenting on the "Hyperledger provided systems", but it seems this is still valid.

File: repository-structure.md (hunk 6 above)
Comment (tkuhrt) [reply repository-structure.md]: That was my impression -- that we were moving all TSC governing process documents to GitHub.

File: repository-structure.md (hunk 4 above)
Comment (shemnon) [reply repository-structure.md]: Adding the SPDX header to reflect the license fo the file does not change the license of a file.  And this is for "Hyperledger" code not third party code.

Comment (arsulegai) [reply repository-structure.md]: Here is an example/file from Hyperledger, but not making use of SPDX header. https://github.com/hyperledger/sawtooth-core/blob/7651cb5552800dbac98559b46339a84eae472ef7/validator/src/lib.rs#L1

Example, having both SPDX and the text as defined in the Apache 2.0 license
https://github.com/hyperledger/besu/blob/ffc22522f8052645ae7453dcc394a70a0f6f797e/besu/src/main/java/org/hyperledger/besu/Besu.java#L1

Example, with only SPDX
https://github.com/hyperledger/fabric/blob/62cd59ca2ee7e37f82ed7e9bc7b08cb9f6822df0/core/peer/peer.go#L4

What I meant here is to accept any of these. Wiki too doesn't mandate one format over the other. Let's discuss pros/cons before putting this as a mandatory requirement here in the common repository structure document.

File: repository-structure.md (hunk 2 above)
Comment (shemnon) [reply repository-structure.md:32]: TSC on 2011-07-15 was no action.

File: repository-structure.md (hunk 4 above)
Comment (shemnon) [reply repository-structure.md]: Per TSC 2021-07-15 I added clarification it applies to new files.

File: repository-structure.md (hunk 5 above)
Comment (shemnon) [reply repository-structure.md:65]: No action for this in this PR.

**Review by tkuhrt — APPROVED** [review]

File: repository-structure.md (hunk 2 above)
Comment (arsulegai) [reply repository-structure.md:32]: Outcome: File is necessary, this file can optionally redirect folks to another location where release notes are maintained.

File: repository-structure.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+Hyperledger projects are required to maintain a standard set of files in each repository. This
+document describes the required and recommended files.
+
+## Required Files with Specified Content
+
+Repositories MUST have these files with the specific content in the linked files, or a file with a
+link to the specified content with minimal exposition. These files MUST be at the root of the
+repository.
+
+-   [`LICENSE`](https://www.apache.org/licenses/LICENSE-2.0.txt) - https://www.apache.org/licenses/LICENSE-2.0.txt\
```
Comment (arsulegai) [inline repository-structure.md:12]: [Just noticed] Do we need to specify it as either `Apache 2.0` or `Creative Commons`?

**Review by arsulegai — APPROVED** [review]

File: repository-structure.md (hunk 8 above)
Comment (shemnon) [reply repository-structure.md:12]: This is intended for code repos and createive commons is not relevant to code.

**Pull request merged**